### PR TITLE
backend authentication resolves #336

### LINKED ISSUE #336: backend authentication

### PR DESCRIPTION
**Pull request by kitzbergerg**
This PR will introduce auth for the backend.

## Todo

- [x] validate tokens
- [x] extract user id from token and allow for passing it on to endpoints
- [x] set config via env
- [x] fix tests (might be quite complicated)

## Basics

- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [ ] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [x] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

File: backend/.env.sample
Comment (markus2330) [inline backend/.env.sample:5]: I think JWKS vs. JWK only leads to unnecessary typos, what about AUTH instead? Explanations what uid:100 could be, see #329:

```suggestion
REMOTE_AUTH_URL=uid;100
```

Comment (kitzbergerg) [reply backend/.env.sample:5]: I think I had AUTH_URL previously, but changed it. The reason is that the URL has only to do with jwks (jsonwebtokens) and I believe REMOTE_AUTH_URL is in that sense less expressive as you would have to first read the code/doc to understand what AUTH means. For somebody that actually has to worry about changing something there JWKS is clearer in my opinion.

**Review by absurd-turtle — APPROVED** [review]
I am not the best judge for Rust code but as far as I can understand the code looks good to me.

File: backend/.env.sample
Comment (markus2330) [reply backend/.env.sample:5]: Ok, I created jwks.permaplant.net (@lukashartl)

**Review by markus2330 — CHANGES_REQUESTED** [review]
Excellent implementation. But please write some documentation how auth works for other devs:

- in unit tests
- for utopia (please also fix existing endpoints)
- how to start the backend with/without auth

**Comment by absurd-turtle** [comment]
I have problems sending request with credentials to the backend. It always displays a cors error.
I think it might be the cors configuration on the server side that is incorrect right now.

Following error is displayed in the frontend:
`Access to XMLHttpRequest at 'http://localhost:8080/api/plants?name=&per_page=30&page=1' from origin 'http://localhost:5173' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

So apparently I am missing the 'Access-Control-Allow-Origin' in the response.
As far as I know this header is set by the server when the Origin sent by the browser matches the url of one of the         .allowed_origin() options.
I can't figure out why this doesn't happen...

I guess we can't use the send_wilcard() option due to this:
> Credential is not supported if the CORS header 'Access-Control-Allow-Origin' is '*'
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials

I haven't been able to figure out the correct cors configuration yet. @kitzbergerg Do you have some ideas how to fix this issue?

**Comment by kitzbergerg** [comment]
> * in unit tests

I'm not quite sure what to do here. The code is quite straight forward. Should I create `doc/tests/backend.md` and make a section there about auth?

> * for utopia (please also fix existing endpoints)

Done

> * how to start the backend with/without auth

Part of the backend README for now. Once everything regarding auth is setup we can move it (but I would to that together with frontend and Keycloak in a separate PR)

**Comment by kitzbergerg** [comment]
@absurd-turtle The setup I have here now works together with the change I made in your frontend branch. Simply follow the steps in the backend README and your frontend stuff should also work.
Note that you either have to have two repos cloned or regularly merge this branch in your frontend branch.

File: backend/README.md
Comment (markus2330) [inline backend/README.md:92]:
```suggestion
Go to <http://localhost:8080/doc/api/swagger/ui/>.  
```

File: backend/README.md
Comment (markus2330) [inline backend/README.md:74]: Do you mean `Client ID` here?

File: backend/README.md
Comment (markus2330) [inline backend/README.md:81]:
```suggestion
Go to `Users` and create a user with username `test`.  
```

File: backend/README.md
Comment (markus2330) [inline backend/README.md:96]: Here I get
```
We are sorry...

Client not found.
```

Did I used the wrong `Client ID`? Furthermore it is unclear if authorizationCode or password should be used.

File: backend/README.md
Comment (markus2330) [inline backend/README.md:19]: Please describe how this URL is built up.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Much better, unfortunately the description does not work for me.

**Comment by markus2330** [comment]
>  I'm not quite sure what to do here. The code is quite straight forward. Should I create doc/tests/backend.md and make a section there about auth?

Yes, a sentence is probably enough (what exactly `init_test_app` does, that it is not necessary to have keycloak running etc.)

**Comment by kitzbergerg** [comment]
@markus2330 I'm currently still changing a bit. I'll include what you mentioned and then rerequest a review.

**Comment by kitzbergerg** [comment]
@markus2330 @absurd-turtle I now removed the conditional compilation. That means as soon as this is merged Keycloak is required and the frontend needs to send tokens to the backend on requests.
The reason is that it is somewhat annoying to set up utoipa to use conditional compilation and we don't need the feature flag anyways as soon as all is merged.

As discussed we can then sit together and merged all of our PRs (frontend, backend and Keycloak) together with @lukashartl at some point.

**Comment by markus2330** [comment]
> As discussed we can then sit together and merged all of our PRs (frontend, backend and Keycloak) together with @lukashartl at some point.

Please vote when the meeting should take place, the deadline for that was on Friday.

**Comment by kitzbergerg** [comment]
> Please vote when the meeting should take place, the deadline for that was on Friday.

I just did. Although I would prefer doing that in a smaller group with just @lukashartl and @absurd-turtle. It's not really necessary to involve more people and we are faster that way.
We can then discuss in some other meeting what has now changed or we just add it as documentation.

File: backend/.env.sample
Comment (lukashartl) [reply backend/.env.sample]: > Ok, I created jwks.permaplant.net (@lukashartl)

I am currently setting up the LXC container for it.

File: backend/.env.sample
Comment (markus2330) [inline backend/.env.sample:7]: We will call it auth.permaplant.net

```suggestion
AUTH_ISSUER_URI=http://localhost:8081/realms/PermaplanT/.well-known/openid-configuration
```

File: backend/README.md
Comment (markus2330) [inline backend/README.md:60]: Please also describe how to test via curl.

**Review by markus2330 — APPROVED** [review]
Few last comments but other than that you can merge this (together with frontend code) today when you see that it actually works today.

File: backend/README.md
Comment (kitzbergerg) [reply backend/README.md:60]: With the setup as described here it is not possible to log in with curl. You would first need to get the access token, but this would require a redirect to Keycloak with a browser based login which is not possible with curl. It would be possible but kind of hacky so I wouldn't suggest it.

Another way of testing would be with Postman, but that is a separate tools so I don't think it fits in the doc.

Comment (markus2330) [reply backend/README.md:60]: Ok, let us forget about curl but we need some solution for our benchmarking/testing tools. So maybe start describing how to do it with httperf

Comment (kitzbergerg) [reply backend/README.md:60]: What I did for httperf is I logged in using Postman and copied the token over. Tokens are valid for 5min so benchmarks can only really run for that long. You can also copy the token from the network tab in the browser after you logged in.

Another way to handle that is to use something called client credentials flow. That way you don't actually need a user that is logged in but instead you can fetch a token from keycloak using the client_id and client_secret pair. That is simpler to do with curl.
However you still have to get a new token every 5min so you have to get a token, run httperf for ~4:30min, fetch a new token, ...
You should also use a separate client for that as the client_secret gets in the way of the normal client.

If you think we need to run benchmarks for more than 5min I can add that somewhere in the doc using curl (but it also involves creating a new client).

Comment (markus2330) [reply backend/README.md:60]: Is there really no way to get tokens valid for longer?

Can you please add what you wrote here in the documentation. Maybe together with how to run benchmarks for the backend.

Comment (kitzbergerg) [reply backend/README.md:60]: I will do that. I would however do that in a separate PR and merge the currently open frontend one first.
Documentation is listed as a TODO in the auth tracking issue: #329

**Pull request merged**
